### PR TITLE
MRG: Error about rank-deficient inputs in Xdawn

### PR DESCRIPTION
**Pull request by larsoner**
I hit a case where `proj` broke `Xdawn`. This should make it work for positive semidefinite instead of only positive definite matrices.

### CODE REVIEW

**Comment by larsoner** [comment]
@alexandrebarachant okay with you?

**Comment by agramfort** [comment]
@alexandrebarachant ok with you?

**Comment by alexandrebarachant** [comment]
I believe the correct way to do this would be to regularize the rank deficient matrices.
switching to eig and selecting the real part will work numerically, but i'm not sure the results will be correct.

**Comment by larsoner** [comment]
I'll check to see whether rank deficiency of `a` or `a` and `b` corrupts the eigenvectors of `eig` or just the last few corresponding to small eigenvalues. Since Xdawn only takes the first few (largest eigenvalue) eigenvectors, if it just breaks the last few that are effectively zero we should be okay I think.

**Comment by larsoner** [comment]
@alexandrebarachant I think you're right. I've changed it just to give a better error message. Can you see if you agree?

File: mne/preprocessing/xdawn.py [hunk 1]
```diff
@@ -176,7 +176,11 @@ def _fit_xdawn(epochs_data, y, n_components, reg=None, signal_cov=None,
         evo_cov = np.matrix(_regularized_covariance(evo, reg))
 
         # Fit spatial filters
-        evals, evecs = linalg.eigh(evo_cov, signal_cov)
+        try:
+            evals, evecs = linalg.eigh(evo_cov, signal_cov)
+        except np.linalg.LinAlgError as exp:
+            raise ValueError('Could not compute eigenvalues, ensure '
+                             'proper regularization (%s)' % (exp,))
```
Comment (jona-sassenhagen) [inline mne/preprocessing/xdawn.py]: Maybe add to the warning where rank deficiencies can come from (e.g. projs)

**Comment by alexandrebarachant** [comment]
looks good. By experience, users don't always know how to add regularization. I think we have a parameter to regularize the covariance estimation in xdawn. it might be good to add something in the error message.

**Comment by agramfort** [comment]
thx @larsoner

**Pull request merged**